src/Symfony/Component/Messenger/Handler/Locator/ContainerHandlerLocator.php

```diff
@@ -30,12 +30,27 @@ public function __construct(ContainerInterface $container)
     public function resolve($message): callable
     {
         $messageClass = \get_class($message);
-        $handlerKey = 'handler.'.$messageClass;
 
-        if (!$this->container->has($handlerKey)) {
+        if (null === $handler = $this->resolveFromClass($messageClass)) {
             throw new NoHandlerForMessageException(sprintf('No handler for message "%s".', $messageClass));
         }
 
-        return $this->container->get($handlerKey);
+        return $handler;
+    }
+
+    private function resolveFromClass($class): ?callable
+    {
+        $handlerKey = 'handler.'.$class;
+        if ($this->container->has($handlerKey)) {
+            return $this->container->get($handlerKey);
+        }
+
+        foreach (class_implements($class, false) as $interface) {
+            if ($handler = $this->resolveFromClass($interface)) {
```

Comment:
not sure this needs to be recursive: `class_implements` already returns all implemented interfaces
also, what about parents?

Reply:
No point of having it recursive, that's true; removed it and added handling of its parents (and the relevant tests, obsly 💃)